### PR TITLE
Ssl wildcard all site types

### LINKED ISSUE #77: 1.2

### LINKED ISSUE #1170: Add ability to add wildcard cert in all site-types
Currently only wildcard cert is used only by `site-wp-command` when site type is `wpsubdom`. We need to add this support in all site types.

### PR DESCRIPTION
**Pull request by kirtangajjar**
Issue ref: https://github.com/EasyEngine/easyengine/issues/1170 and https://github.com/EasyEngine/site-command/issues/77

Other than that, this PR  - 
1. corrects SSL behaviour as discussed with team(use --ssl and --wildcard instead of --le).
2. Adds WWW domain while requesting cert - https://github.com/EasyEngine/easyengine/pull/1171/commits/c57626365d16e2f3d863c9a6c8adae92a6b849c2
3. Adds function to reload ee-nginx-proxy container - https://github.com/EasyEngine/easyengine/pull/1171/commits/192fdc45a18ab116834e8dd74ee9ecb4586f270a
4. Fix http challenge creation - https://github.com/EasyEngine/easyengine/pull/1171/commits/6977a1df104fbcf4f788541d51e8d07d5ccc3314
5. Add ability to inherit cert - https://github.com/EasyEngine/easyengine/pull/1171/commits/f74c8e199030ba9103e3294558a67afa7db29398

Related PRs:
https://github.com/EasyEngine/site-command/pull/98
https://github.com/EasyEngine/site-wp-command/pull/17
https://github.com/EasyEngine/shell-command/pull/5

closes #1170 and closes #77

### CODE REVIEW

File: php/class-ee-site.php [hunk 1]
```diff
@@ -355,35 +422,72 @@ private function reload_services( $services, $reload_commands ) {
...
+	 * Returns all domains required by cert
+	 *
+	 * @param string $site_name Name of site
+	 * @param $wildcard  Wildcard cert required?
+	 *
+	 * @return array
+	 */
+	private function get_cert_domains( string $site_name, $wildcard ) : array {
+		$domains = [ $site_name ];
+		$has_www = strpos( $site_name, 'www.' ) === 0;
```
Comment (rahulsprajapati) [inline php/class-ee-site.php]: > $has_www = ( strpos( $site_name, 'www.' ) === 0 );

File: php/class-ee-site.php [hunk 2]
```diff
@@ -355,35 +422,72 @@ private function reload_services( $services, $reload_commands ) {
...
+	/**
+	 * If the domain has www in it, returns a domain without www in it.
+	 * Else returns a domain with www in it.
+	 *
+	 * @param string $site_name Name of site
+	 *
+	 * @return string Domain name with or without www
+	 */
+	private function get_www_domain( string $site_name ) : string {
+		$has_www = strpos( $site_name, 'www.' ) === 0;
```
Comment (rahulsprajapati) [inline php/class-ee-site.php]: > $has_www = ( strpos( $site_name, 'www.' ) === 0 );

File: php/site-utils.php [hunk 3]
```diff
@@ -143,27 +143,55 @@ function setup_site_network( $site_name ) {
...
  * Adds www to non-www redirection to site
  *
- * @param string $site_name Name of the site.
- * @param bool $le          Specifying if letsencrypt is enabled or not.
+ * @param string $site_name name of the site.
+ * @param bool $ssl         enable ssl or not.
+ * @param bool $inherit     inherit cert or not.
  */
-function add_site_redirects( $site_name, $le ) {
+function add_site_redirects( string $site_name, $ssl, $inherit ) {
```
Comment (rahulsprajapati) [inline php/site-utils.php]: Please maintain code consistency. here you're defining data type of function variable and other places you're not. Either you define it for all or not at all.
ref: https://github.com/EasyEngine/easyengine/pull/1171/files#diff-43605c4fabcca6b901ccbab377546cfeR404

File: php/site-utils.php [hunk 4]
```diff
@@ -143,27 +143,55 @@ function setup_site_network( $site_name ) {
...
 
 	if ( $has_www ) {
 		$site_name_without_www = ltrim( $site_name, '.www' );
 		// ee site create www.example.com --le
-		if ( $le ) {
+		if ( $ssl ) {
 			$content = "
 server {
 	listen  80;
+	server_name  $site_name_without_www;
```
Comment (rahulsprajapati) [inline php/site-utils.php]: Can this go in mustache template? It will be easy to read and update config in that.

File: php/site-utils.php [hunk 5]
```diff
@@ -179,12 +207,25 @@ function add_site_redirects( $site_name, $le ) {
 	} else {
 		$site_name_with_www = 'www.' . $site_name;
 		// ee site create example.com --le
-		if ( $le ) {
+		if ( $ssl ) {
 
 			$content = "
 server {
 	listen  80;
+	server_name  $site_name_with_www;
```
Comment (rahulsprajapati) [inline php/site-utils.php]: Can this go in mustache template? It will be easy to read and update config in that.

**Comment by rahulsprajapati** [comment]
@mbtamuli Please review this PR for config related changes.
ref: https://github.com/EasyEngine/easyengine/pull/1171/files#diff-6dd1e8dbaffdb03759d37d087585f8a2R221

**Comment by mbtamuli** [comment]
@rahulsprajapati I've checked all the config regarding this wildcard configuration.

File: php/class-ee-site.php [hunk 6]
```diff
@@ -347,6 +359,61 @@ private function reload_services( $services, $reload_commands ) {
...
+	 * @param string $site_name      Name of the site for ssl.
+	 * @param string $needs_wildcard Does site needs wildcard.
+	 *
+	 * @throws Exception
+	 */
+	protected function inherit_certs( $site_name, $needs_wildcard ) {
+		$parent_site_name = implode( '.', array_slice( explode( '.', $site_name ), 1 ) );
+		$parent_site      = Site::find( $parent_site_name, [ 'site_ssl', 'site_ssl_wildcard' ] );
+
+		if ( $needs_wildcard ) {
```
[the remark this answers is not in the document]
Comment (kirtangajjar) [reply php/class-ee-site.php]: Done

File: php/class-ee-site.php (hunk 1 above)
Comment (kirtangajjar) [reply php/class-ee-site.php]: Resolved.

File: php/class-ee-site.php (hunk 2 above)
Comment (kirtangajjar) [reply php/class-ee-site.php]: Resolved.

File: php/site-utils.php (hunk 3 above)
Comment (kirtangajjar) [reply php/site-utils.php]: Resolved.

File: php/site-utils.php (hunk 5 above)
Comment (kirtangajjar) [reply php/site-utils.php]: Resolved.

File: php/site-utils.php [hunk 7]
```diff
@@ -143,61 +143,48 @@ function setup_site_network( $site_name ) {
...
-	listen  80;
-	server_name  $site_name_with_www;
-	return  301 http://$site_name\$request_uri;
-}";
-		}
+		$server_name = 'www.' . $site_name;
 	}
+
+	$conf_data = [
+		'site_name' => $site_name,
```
Comment (rahulsprajapati) [inline php/site-utils.php]: alignment.

Comment (kirtangajjar) [reply php/site-utils.php]: Resolved.

**Comment by renrayut** [comment]
HI , when i try ee site create example.com --ssl=le get ssl file (.crt, .key,. pem ) on easyengine/nginx-proxy:v4.0.0-beta.6 container . it default type RSA-4096 . but i need type RSA-2048 for site . how to do?
cos i try to convert by openssl still can't complete. 
thanks for reply .

**Pull request merged**